alerting/src/main/kotlin/org/opensearch/alerting/alerts/AlertMover.kt

```diff
@@ -39,196 +38,209 @@ import org.opensearch.index.query.QueryBuilders
-                            return
-                        alertIndex = monitor.dataSources.alertsIndex
-                        alertHistoryIndex =
-                            if (monitor.dataSources.alertsHistoryIndex == null) alertHistoryIndex
-                            else monitor.dataSources.alertsHistoryIndex!!
+        /**
+         * Moves defunct active alerts to the alert history index when the corresponding workflow or trigger is deleted.
+         *
+         * The logic for moving alerts consists of:
+         * 1. Find active alerts:
+         *      a. matching workflowId if no workflow is provided (postDelete)
+         *      b. matching workflowid and chained alert triggerIds if monitor is provided (postIndex)
+         * 2. Move alerts over to DataSources.alertsHistoryIndex as DELETED
+         * 3. Delete alerts from monitor's DataSources.alertsIndex
+         * 4. Schedule a retry if there were any failures
+         */
+        suspend fun moveAlerts(client: Client, workflowId: String, workflow: Workflow?, monitorCtx: MonitorRunnerExecutionContext) {
+            var alertIndex = ALERT_INDEX
+            var alertHistoryIndex = ALERT_HISTORY_WRITE_INDEX
+            if (workflow != null) {
+                if (
+                    workflow.inputs.isNotEmpty() && workflow.inputs[0] is CompositeInput &&
+                    (workflow.inputs[0] as CompositeInput).sequence.delegates.isNotEmpty()
+                ) {
+                    val i = 0
+                    val delegates = (workflow.inputs[i] as CompositeInput).sequence.delegates
+                    try {
+                        var getResponse: GetResponse? = null
+                        while (i < delegates.size && (getResponse == null || getResponse.isExists == false)) {
+                            getResponse =
+                                client.suspendUntil {
+                                    client.get(
+                                        GetRequest(ScheduledJob.SCHEDULED_JOBS_INDEX, delegates[0].monitorId),
+                                        it
+                                    )
+                                }
+                            if (getResponse!!.isExists) {
+                                val monitor =
+                                    ScheduledJobUtils.parseMonitorFromScheduledJobDocSource(
+                                        monitorCtx.xContentRegistry!!,
+                                        response = getResponse
+                                    )
+                                /** check if alert index is initialized **/
+                                if (monitorCtx.alertIndices!!.isAlertInitialized(monitor.dataSources) == false)
+                                    return
+                                alertIndex = monitor.dataSources.alertsIndex
+                                alertHistoryIndex =
```

Comment:
What if multiple monitors have different data sources and part of the same workflow? This wont move all the alerts and they get stuck defunct.

Reply:
Also this is only for alerts that are not going to be in the audit state, right?

Reply:
Workflows currently assume same datasources for all monitors

Reply:
Does the workflow enforce a single data source for all monitors?

Reply:
we are not doing a validation but right now there is no real world scenario where we add monitors with different data sources because-
rest apis dont support data sources and are populatd with alerting system indices
security analytics detectos have one data source each and wont intermix

Reply:
Please add a github issue to ensure workflows have monitors of a single data source only